system/src/Grav/Common/Page/Medium/ImageMedium.php

```diff
@@ -232,31 +218,36 @@ public function srcset($reset = true)
      * @return $this
      */
     public function derivatives($min_width, $max_width, $step = 200) {
-      $width = $min_width;
-
-      // Do not upscale images.
-      if ($max_width > $this->get('width')) {
-        $max_width = $this->get('width');
-      }
-
-      while ($width <= $max_width) {
-        $ratio = $width / $this->get('width');
-        $derivative = MediumFactory::scaledFromMedium($this, 1, $ratio);
-        if (is_array($derivative)) {
-          $this->addDerivative($derivative['file']);
+        $width = $min_width;
+
+        // Do not upscale images.
+        if ($max_width > $this->get('width')) {
+            $max_width = $this->get('width');
         }
-        $width += $step;
-      }
-      return $this;
-    }
 
-    /**
-     * Add a derivative
-     *
-     * @param  ImageMedium $image
-     */
-    public function addDerivative(ImageMedium $image) {
-      $this->derivatives[$image->url()] = $image->get('width');
+        // Clear any alternatives that have already been implied by the image's
+        // filename (eg. ones that had "@2x" or similar appended to their
+        // filenames)
+        $this->clearAlternatives();
```

Comment:
This highlights an issue with the `derivatives` method. When using it in Twig templates, any filters applied before I call it won't affect the derivatives. It would of course be best if the order in which I called the methods didn't matter, but maybe we ought to call `$this->reset()` here to make sure the original medium and the derivatives at least look the same when rendered?